### PR TITLE
Pr/csr initials

### PR DESCRIPTION
**Pull request by veryprofessionaldodo**
| - | - |
| --- | --- |
| Dependencies | [ThreeJS](https://github.com/mrdoob/three.js), [POSTPROCESS](https://github.com/vanruesc/postprocessing)
| Decisions | There are several limitations to the Pre-Render Compose strategy, so a Client-Side Rendering strategy was implemented. Using the ThreeJS graphics engine, this branch integrates real-time rendering with the Ripe SDK as an alternative to the classic pre-render approach. Currently, there is support for changing parts using setParts, orbital rotation around the mesh, crossfades, several easing strategies for rotations and highlights, rotation locking to horizontal or vertical axis, initials support, some postprocessing features such as bloom, multi-animation support in a glTF file, mouse drift and more. |
| Animated GIF | full3 |

### CODE REVIEW

File: src/python/ripe_demo/static/js/main.js [hunk 1]
```diff
@@ -396,7 +396,13 @@ window.onload = function () {
                     materialEasing: "easeInOutQuad",
                     crossfadeEasing: "easeInOutQuad",
                     initialsPlacement: "center",
-                    introAnimation: "SlideIn",
+                    introAnimation: "camera_front_pan_1.fbx",
+                    introAnimation: "camera_front_pan_1.glb",
+                    introAnimation: "mesh_slide_in_1.glb",
+                    //introAnimation: "mesh_slide_in_1.fbx",
```
Comment (joamag) [inline src/python/ripe_demo/static/js/main.js:404]: no commented code

**Review by joamag — CHANGES_REQUESTED** [review]

File: src/python/ripe_demo/static/js/main.js (hunk 1 above)
Comment (veryprofessionaldodo) [reply src/python/ripe_demo/static/js/main.js]: Fixed!

**Comment by joamag** [comment]
@veryprofessionaldodo don't forget to run `npm run pretty` before any commit

**Comment by joamag** [comment]
@veryprofessionaldodo Other notes:
* `var` is not allowed, please use `const` or `let`
* Avoid committing code that is commented out
* Document the expected data types for parameters and return values - avoid `{*}` in docstrings
* Start inline comments with lowercase letters

**Comment by joamag** [comment]
@veryprofessionaldodo more notes
* Use `for of` for simple iterator driven iteration instead of using index based iteration
```javascript
    for (let i = 0; i < this.raycastingMeshes.length; i++) {
        await this.assetManager.disposeMesh(this.raycastingMeshes[i]);
    }
```

becomes

```javascript
    for (const raycastingMesh of this.raycastingMeshes) {
        await this.assetManager.disposeMesh(raycastingMesh);
    }
```

**Comment by joamag** [comment]
@veryprofessionaldodo more notes
* Use always `"` instead of `'`
* Do not commit `TODO`, if not implemented prefer the `NotImplementedException`

**Comment by joamag** [comment]
@veryprofessionaldodo one more
* Avoid adding large blobs to the repo as that increases the overall repo size

**Comment by joamag** [comment]
@veryprofessionaldodo this branch is now locked if you need to do some quick fixes regarding visual issues please fork this branch

**Comment by joamag** [comment]
Closing this PR for now will re-open once the branch is more stable

**Comment by veryprofessionaldodo** [comment]
@joamag Just saw that I got assigned to this, but this is branch is more than half a year out of date compared to the [sandbox version](https://github.com/ripe-tech/sandbox/tree/pr/ripe-csr). I think that this should be closed, and a new, more fleshed out, more correct version should be made from that version.

**Comment by veryprofessionaldodo** [comment]
@joamag Should this PR be closed? It's quite out of date as I wrote before, with the new [sandbox version](https://github.com/ripe-tech/sandbox/tree/pr/ripe-csr) having almost a year more of updates. What do you think?

**Comment by joamag** [comment]
> @joamag Should this PR be closed? It's quite out of date as I wrote before, with the new [sandbox version](https://github.com/ripe-tech/sandbox/tree/pr/ripe-csr) having almost a year more of updates. What do you think?

We'll probably have to start with this API and then refactor and add new features that have been developed on top of this.

**Comment by veryprofessionaldodo** [comment]
@joamag I strongly suggest against using this PR for completing CSR, for a lot of different reasons. The repo I [mentioned before](https://github.com/ripe-tech/sandbox/tree/pr/ripe-csr) has these changes (and more that I can't remember):
- is already completely self-contained, and serves only as an optional extension of the SDK, just like requested, with a [PR](https://github.com/ripe-tech/ripe-sdk/pull/318) done to support this;
- features the capability to generate masked images for RIPE-Compose (a lot of changes were done for this);
- much higher resilience for random errors;
- capability to highlight multiple meshes using the same hover;
- complete compliance with latest version of CSR specification, which means it will work using the latest updates done to the editor. This includes all the capabilities allowed by the latest specification, including part grouping and smart defaults for meshes not in the build;
- improved compatibility with current events from RIPE SDK, including config;
- major refactors to streamline texture loading;
- improvements to how views are handled;
- major improvements to asynchronous processing of events for better performance and time-to-first-frame;

And all of this is done while the code simultaneously underwent major refactors for improved code structure, and a huge number of bug fixes. If I had to guess, between 50 and 80 bugs were solved in the new version. 

I suggest using the other branch for CSR. We currently have under 7 weeks until the end of Q1, and there's still more work to be done outside of CSR to get this in a production-ready state. I don't think we have the time to backport all the features and fixes to this outdated branch, while still working on the other requirements to get this running in time.

**Pull request closed without merging**